documentation/source/building.rst

```diff
@@ -222,6 +222,31 @@ GPI
         and loading from libraries that `aren't` prefixed with "lib".
         Paths `should not` contain commas.
 
+PyGPI
+-----
+
+.. envvar:: COCOTB_ENTRY_POINT
+
+    The Python module and callable that starts up the Python cosimulation environment.
+    This defaults to :data:`cocotb:_initialise_testbench`, which is the cocotb standard entry point.
+    User overloads can be used to enter alternative Python frameworks or to hook existing cocotb functionality.
+    The variable is formatted as ``path.to.entry.module:entry_point.function``.
+    The string before the colon is the Python module to import
+    and the string following the colon is the object to call as the entry function.
+
+    The entry function must be a callable matching this form:
+
+    * ``entry_function(argv: List[str]) -> bool``
+
+    The entry module must have the following additional functions defined:
+
+    * ``_sim_event(level: int) -> bool``
+    * ``_log_from_c(logger_name: str, level: int, filename: str, lineno: int, msg: str, function_name: str)) -> bool``
+    * ``_filter_from_c(logger_name: str, level: int) -> bool``
+
+    The entry module functions return ``True`` on success. On failure, they should return ``False`` or raise an :class:`Exception`.
```

Comment:
This API seems odd to me; I'd prefer the API be something like
```
def my_handle_event(level: int) -> None:
   # raise exception on error

def my_log_from_c(...) -> None:
   # raise exception on error

def my_filter_from_c(...) -> bool:
   # return bool indicating whether to filter the message, raise exception on error

def entry_function(argv: List[str]) -> bool:
    # raise exception on error
    cocotb.simulator.set_sim_event_handler(my_handle_event)
    cocotb.simulator.set_log_handler(my_log_from_c, my_filter_from_c)
```
or perhaps
```
def entry_function(argv: List[str]):
    return (my_handle_event, my_log_from_c, my_filter_from_c)
```

Returning `bool`s to indicate errors is unpythonic, and I don't think we should be doing it.

Reply:
Note that this approach allows different handlers to be registered based on the `argv`.

Reply:
The interface is odd, but I just encoded what cocotb was already doing. I'd prefer the first approach, and it's where I intend to end up; which is why I mentioned "these additional requirements will be relaxed in the future". Doing it all at once will make this diff huge since the object will need to live in some common library used by both the simulatormodule and gpi_embed. I have mentioned maybe we can combine gpi_embed and simulatormodule into a single library to avoid that unnecessary complexity.

Reply:
I know it's better, I'm trying to chip away bit by bit because large diffs don't get reviewed.

Reply:
We can change them to return None and use Exceptions. I noticed that `_initialise_testbench` returned True. Whichever way it goes make the functions we have match the documented behavior.

Reply:
For now let's change them to raise on error and fixup the documentation. As Kaleb said we can iterate and improve on it in a later PR.

Reply:
I removed the boolean returns. But the names remain the same and they remain required-by-name members of the init module. A follow on will move the interface to instead be registered by calling functions on the `simulator` module.